Add automatic download of bioenergy potentials and cost

timtroendle opened the pull request:
* Add automatic download, Fixes #194.
* Update default scenario years to align them, Fixes #195.
* Fix cost of municipal solid waste, Fixes #193.

## Checklist

Any checks which are not relevant to the PR can be pre-checked by the PR creator. All others should be checked by the reviewer. You can add extra checklist items here if required by the PR.

- [x] CHANGELOG updated
- [x] Minimal workflow tests pass
- [x] Tests added to cover contribution
- [x] Documentation updated
- [x] Configuration schema updated

brynpickering commented on scripts/biofuels.py:
```diff
@@ -68,81 +59,95 @@ class GlobCover(Enum):
         f_cost.write(str(weighted_cost))
 
 
 def allocate_potentials(national_potentials, units, population, land_cover):
```
I suspect all of this would be best done in pandas, rather than xarray:

```
shares = pd.concat([population, land_cover[FOREST].sum(axis=1), land_cover[FARM].sum(axis=1)], axis=1, keys=['population_share', 'forest_share', "farm_share"]).groupby(units.country_code).transform(lambda x: x / x.sum())
regional_potentials = pd.merge(national_potentials.reset_index(), units["country_code"].reset_index(), on="country_code").set_index(["id", "feedstock"]).unstack("feedstock")
regional_potentials_sum = regional_potentials.groupby(PROXIES, axis=1).sum()
regional_potentials_weighted = regional_potentials_sum.mul(shares).sum(axis=1)
```

timtroendle replied:
Maybe. I gave this a try, but it didn't work right away. A pure Pandas approach would require more changes, as the inputs are xarray objects.

brynpickering commented on scripts/biofuels.py:
```diff
@@ -68,81 +59,95 @@ class GlobCover(Enum):
-    weighted_cost = (costs * national_potentials / national_potentials.sum().sum()).sum().sum()
-    with open(path_to_costs, "w") as f_cost:
+    total_potential.to_csv(paths_to_output.potentials, index=True, header=True)
+    weighted_cost = (
```
If we have costs per region, why not keep them as regional values? They'll be the same for every subnational region, but will be different between countries.

timtroendle replied:
Thought about the same when implementing. OK, but in a separate issue #200.

brynpickering reviewed (changes requested):
Looks fine. Some comments to address and perhaps some new issues to create:

1. PROXIES -> config
2. retain regional distribution of costs
3. verify that your PROXIES for subnational regionalisation are similar to the NUTS2 data given in the Excel

timtroendle commented on scripts/biofuels.py:
```diff
@@ -1,22 +1,13 @@
-    "high": "High availability scenario"
-}
-
 PROXIES = {
```
Yes!

timtroendle commented on scripts/biofuels.py:
```diff
@@ -68,81 +59,95 @@ class GlobCover(Enum):
-           .rename(path.stem) / GJ_TO_MWH
-         for path in paths_to_costs],
-        axis=1
+    national_costs = (
```
Hm, most tech costs are given in Eur2013, so I'd assume both are equally good.

timtroendle commented on scripts/biofuels/extract_biofuels.py:
```diff
@@ -0,0 +1,79 @@
+    df[COL_NAME_FEEDSTOCK] = df[COL_NAME_FEEDSTOCK].map(FEEDSTOCK_NAME_MAP)
+    df[COL_NAME_COUNTRY] = df[COL_NAME_COUNTRY].map(utils.eu_country_code_to_iso3)
+    df.rename(columns=COL_NAME_MAP, inplace=True)
+    df.value = pd.to_numeric(df.value, errors='coerce')
```
Ok but is there a specific reason?

timtroendle commented on scripts/biofuels/extract_biofuels.py:
```diff
@@ -0,0 +1,79 @@
+    # Renames countries to ISO3.
+    # Renames column names
+    # Sets 1 missing value (BIH med potential 2010) to 0.
+    # Sets 1 missing value (NOR, high cost 2030) to 0.
```
ok. I am using another scenario's value as values barely change between scenarios (and are tiny anyways)

timtroendle commented on scripts/biofuels/extract_biofuels.py:
```diff
@@ -0,0 +1,79 @@
+    # Renames scenarios and feedstocks.
+    # Renames countries to ISO3.
+    # Renames column names
+    # Sets 1 missing value (BIH med potential 2010) to 0.
```
ok. But both are 0.

timtroendle commented:
Ok. I implemented most of your recommendations and I moved PROXIES (and actually much more) to the config. The other two points should be handled in other issues in my opinion (I opened #200).

brynpickering commented on scripts/biofuels.py:
```diff
@@ -68,81 +45,97 @@ class GlobCover(Enum):
+        .set_index(["id", "feedstock"])
+        .to_xarray()
+    )
+    ds["population"] = (
+        population
+        .groupby(units.country_code)
+        .transform(lambda x: x / x.sum())
+        .rename("population")
+    )
+    ds["forest"] = (
+        land_cover[FOREST]
+        .sum(axis=1)
+        .groupby(units.country_code)
+        .transform(lambda x: x / x.sum())
+        .rename("forest")
+    )
+    ds["farmland"] = (
+        land_cover[FARM]
+        .sum(axis=1)
+        .groupby(units.country_code)
+        .transform(lambda x: x / x.sum())
+        .rename("farmland")
+    )
```
```suggestion
    shares = (
        pd.concat(
            [population, land_cover[FOREST].sum(axis=1), land_cover[FARM].sum(axis=1)], 
            axis=1, 
            keys=['population', 'forest', "farmland"]
        )
        .groupby(units.country_code)
        .transform(lambda x: x / x.sum())
    )
```

brynpickering commented on scripts/biofuels.py:
```diff
@@ -68,81 +45,97 @@ class GlobCover(Enum):
+        .transform(lambda x: x / x.sum())
+        .rename("farmland")
+    )
+    ds["proxy"] = pd.Series(proxies).rename("proxy").rename_axis(index="feedstock").to_xarray()
+    proxy_value = xr.ones_like(ds.value)
+    for feedstock in ds.feedstock:
+        for id in ds.id:
+            proxy = ds.sel(id=id, feedstock=feedstock).proxy.item()
+            proxy_value.loc[dict(id=id, feedstock=feedstock)] = ds.sel(id=id)[proxy].item()
+    ds["value"] = ds.value * proxy_value
+    return ds.value.sum("feedstock").rename(NAME).to_series()
```
```suggestion
    regional_potentials_sum = regional_potentials.groupby(proxies, axis=1).sum()
    regional_potentials_weighted = regional_potentials_sum.mul(shares).sum(axis=1)
    return regional_potentials_weighted.rename(NAME)
```

brynpickering commented on scripts/biofuels.py:
```diff
@@ -68,81 +45,97 @@ class GlobCover(Enum):
+        proxies=proxies
+    )
+    total_potential.to_csv(paths_to_output.potentials, index=True, header=True)
+    weighted_cost = (
+        (national_costs * national_potentials / national_potentials.sum(["country_code", "feedstock"]))
+        .sum(["country_code", "feedstock"])
+        .item()
     )
```
```suggestion
    weighted_cost = (
          national_costs
          .mul(national_potentials)
          .div(national_potentials.sum())
          .sum()
    )
```

brynpickering replied:
and move this above `if (len(units.index) == 1) and (units.index[0] == "EUR"):`

brynpickering commented on scripts/biofuels.py:
```diff
@@ -68,81 +45,97 @@ class GlobCover(Enum):
+            .sum(["country_code", "feedstock"])
+            .expand_dims({"country_code": ["EUR"]})
+        )
+        national_potentials = national_potentials.sum("country_code").expand_dims({"country_code": ["EUR"]})
```
```suggestion
        national_potentials = (
            national_potentials
            .rename(lambda x: "EUR", level="country_code")
            .sum(level=["country_code", "feedstock"])
        )
```

brynpickering commented on scripts/biofuels.py:
```diff
@@ -68,81 +45,97 @@ class GlobCover(Enum):
     if (len(units.index) == 1) and (units.index[0] == "EUR"): # special case for continental level
-        national_potentials = pd.DataFrame(index=["EUR"], data=national_potentials.sum(axis=0).to_dict())
-        costs = pd.DataFrame(index=["EUR"], data=costs.mean(axis=0).to_dict())
+        national_costs = (
+            (national_costs * national_potentials / national_potentials.sum(["country_code", "feedstock"]))
+            .sum(["country_code", "feedstock"])
+            .expand_dims({"country_code": ["EUR"]})
+        )
```
```suggestion
```

In preference of bringing 'weighted_costs' up from below, which is a duplication of this if undertaken before changing 'national_potentials'.

brynpickering reviewed (changes requested):
OK, I made suggestions to have everything remain in pandas the whole way through (rather than the current approach of pandas -> xarray -> pandas). I think it's cleaner and also quicker.

It's my first attempt at using 'suggestions' and I think it ultimately made it more confusing than I had hoped. Sorry. Let me know if you'd prefer that I just push a commit with the necessary updates.

On another note. Should `biofuels.py` be in the biofuels subdir?

timtroendle commented:
It's indeed a little complicated. Plus, you couldn't know whether it worked (it didn't), you had no code autocorrect, and you appear as commiter only, not as author. So my conclusion would be: worth the try, but not a terribly good solution. :P

Thanks in any case. This is cleaner now. Will merge once you give your ok.

brynpickering commented:
Haha, ok, I won't try making such lengthy suggestions again ;)

> On another note. Should biofuels.py be in the biofuels subdir?

Did you consider this?

timtroendle commented:
Yes, I did. Sorry, I was running the minimal workflow first before pushing. Now it's updated.
